lib/evaluate/intrinsics.cc

```diff
@@ -0,0 +1,1244 @@
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "intrinsics.h"
+#include "expression.h"
+#include "type.h"
+#include "../common/enum-set.h"
+#include "../common/fortran.h"
+#include "../common/idioms.h"
+#include <map>
+#include <ostream>
+#include <sstream>
+#include <string>
+#include <utility>
+
+using namespace Fortran::parser::literals;
+
+namespace Fortran::evaluate {
+
+using common::TypeCategory;
+
+// This file defines the supported intrinsic procedures and implements
+// their recognition and validation.  It is largely table-driven.  See
+// documentation/intrinsics.md and section 16 of the Fortran 2018 standard
+// for full details on each of the intrinsics.  Be advised, they have
+// complicated details, and the design of these tables has to accommodate
+// that complexity.
+
+// Dummy arguments to generic intrinsic procedures are each specified by
+// their keyword name (rarely used, but always defined), allowable type
+// categories, a kind pattern, a rank pattern, and information about
+// optionality and defaults.  The kind and rank patterns are represented
+// here with code values that are significant to the matching/validation engine.
+
+// These are small bit-sets of type category enumerators.
+// Note that typeless (BOZ literal) values don't have a distinct type category.
+// These typeless arguments are represented in the tables as if they were
+// INTEGER with a special "typeless" kind code.  Arguments of intrinsic types
+// that can also be be typeless values are encoded with an "elementalOrBOZ"
+// rank pattern.
+using CategorySet = common::EnumSet<TypeCategory, 8>;
+static constexpr CategorySet Int{TypeCategory::Integer};
+static constexpr CategorySet Real{TypeCategory::Real};
+static constexpr CategorySet Complex{TypeCategory::Complex};
+static constexpr CategorySet Char{TypeCategory::Character};
+static constexpr CategorySet Logical{TypeCategory::Logical};
+static constexpr CategorySet IntOrReal{Int | Real};
+static constexpr CategorySet Floating{Real | Complex};
+static constexpr CategorySet Numeric{Int | Real | Complex};
+static constexpr CategorySet Relatable{Int | Real | Char};
+static constexpr CategorySet IntrinsicType{
```

Comment:
Consider just `Intrinsic` for the name of this set. The ones above don't have `Type` in their names.

`AnyType` below seems okay because it is special (and just `Any` wouldn't be specific enough).

Reply:
I use `IntrinsicType` to avoid confusion in a module that's concerned with intrinsic functions.

It might be better to append `Type` to the names of the others.  Or rename `IntrinsicType` to `AnyIntrinsic` or `AnyIntrinsicType`.  Which sounds better to you?

Reply:
I think I prefer adding `Type` to the others.